Implement spatial hadamards in the fixed bulk

BSchelpe opened the pull request:
This will close issue #748. It is currently a draft PR, because it still needs more work doing (most obviously the left-right hadamard implementing) and testing. I am pushing it now, so that Gilad can do testing on top of it.

BSchelpe committed:
First version of spatial hadamards in the fixed bulk

BSchelpe committed:
Merge branch 'main' into issue748
Updating with main.

BSchelpe committed:
Fix ExtendedPlaquetteCollection

BSchelpe committed:
Small fixes

BSchelpe committed:
more small fixes

BSchelpe committed:
Fix plaquette vs rpng descriptions

BSchelpe committed:
Merge branch 'main' into issue748

BSchelpe committed:
Merge branch 'main' into issue748

BSchelpe committed:
small change

BSchelpe committed:
Correct how sbb is pased through to tlb

BSchelpe committed:
Merge branch 'issue748' of https://github.com/BSchelpe/tqec into issue748
merge with updates

BSchelpe committed:
Merge branch 'main' into issue748

BSchelpe committed:
Merge branch 'main' into issue748

BSchelpe committed:
Merge branch 'main' into issue748

BSchelpe committed:
Remove some corner plaquettes to avoid interference between arms

BSchelpe committed:
Merge branch 'main' into issue748

BSchelpe committed:
Merge branch 'main' into issue748

BSchelpe committed:
Merge branch 'main' into issue748

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -862,14 +879,15 @@ def get_spatial_cube_arm_raw_template(self, arms: SpatialArms) -> RectangularTem
             raise TQECError(f"Unrecognized spatial arm(s): {arms}.")
 
-    def get_spatial_cube_arm_rpng_descriptions(
+    def get_spatial_cube_arm_plaquettes(
```
Does renaming the function break any tests or external use cases

BSchelpe replied:
Lourens has very kindly offered to do this.

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1031,23 +1008,97 @@ def _get_left_right_spatial_cube_arm_rpng_descriptions(
+            del mapping[up]  # mapping[up] = _corners[1]
+        # Replaces the bottom plaquette if there is an arm, which might modify the plaquette to a
+        # corner, or extended triangle.
         if SpatialArms.LEFT in arms and _sbb == Basis.X and SpatialArms.DOWN in v.spatial_arms:
```
Maybe combine these `if`s into a single one with an `or`?

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1106,6 +1158,108 @@ def _get_up_down_spatial_cube_arm_rpng_descriptions(
+        such as the lattice surgery implementation of a ``CZ`` gate.
+
+        Warning:
+        This method is tightly coupled with
```
If they are tightly coupled, should they really be different methods?

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1384,3 +1538,445 @@ def get_spatial_horizontal_hadamard_plaquettes(
+    #                Extended stabiliser Hadamards                #
+    ###############################################################
+
+    def get_spatial_z_above_rght_and_2_arm_extended_stabiliser_hadamard_raw_template(
```
Would it be better to have one `get_spacial_z_above_right` and provide additional parameters (like the template)?

smburdick reviewed (commented):
* Overall, many of the new functions  contain a lot of the same code. I think they can be generalized, to make the code easier to maintain and understand.
* Please update the testing modules to include your code changes, and provide some use cases.

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1031,23 +1008,97 @@ def _get_left_right_spatial_cube_arm_rpng_descriptions(
+        reset: Basis | None = None,
+        measurement: Basis | None = None,
+    ) -> Plaquettes:
+        raise NotImplementedError()
```
I think it's better to raise a specific issue or have a checklist item to implement this, in order to raise more visibility, rather than making the code as-is more complex.

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1384,3 +1538,445 @@ def get_spatial_horizontal_hadamard_plaquettes(
+        tlb, otb = top_left_basis, top_left_basis.flipped()
+        # Generating plaquette descriptions we will need later.
+        extended_plaquette_collection = self.get_extended_plaquettes(reset, measurement)
+        bulk = dict()
```
It's just a little cleaner than declaring an empty dict and assigning each member imperatively.

BSchelpe replied:
Thanks for the explanation. Happy to change it.

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -923,72 +943,29 @@ def get_spatial_cube_arm_rpng_descriptions(
-            return self._get_left_right_spatial_cube_arm_rpng_descriptions(
-                spatial_boundary_basis, arms, linked_cubes, reset, measurement
-            )
+            if is_hadamard:
```
I was just noticing that the new functions look similar and could possibly be combined, for the sake of easier maintainability, like if we decide to have more than 2 arms in the future.

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -923,72 +943,29 @@ def get_spatial_cube_arm_rpng_descriptions(
+                )
         raise TQECError(f"Got an invalid arm: {arms}.")
 
-    def get_spatial_cube_arm_plaquettes(
```
ok.

BSchelpe committed:
Merge branch 'main' into issue748

BSchelpe committed:
Small changes from smburdick's review

BSchelpe committed:
Merge branch 'main' into issue748

BSchelpe committed:
Merge branch 'main' into issue748

BSchelpe committed:
Expand logic of if else to cover other conigurations

BSchelpe committed:
Merge branch 'main' into issue748

BSchelpe commented:
After running more test configurations, I realised that the above diagrams do not fully capture all possible configurations of the hadamard pipe, in particular situations where both u and v have arms beyond their connecting arm. Below are more comprehensive (also more schematic) diagrams of the possible junctions. The PR logic has also been updated to reflect this. 
Note that in the 'horseshoe' configurations (ie the first in each case) it is a full-sized rectangle added to the boundary of the connecting pipe, whereas in the 'single pipe' (4th) configurations they are half-width rectangles.
![IMG_3200](https://github.com/user-attachments/assets/bd8c264c-a0a1-45e5-a662-18de58976692)
![IMG_3199](https://github.com/user-attachments/assets/ef22e229-0e39-469f-b777-19002fb3a62e)

BSchelpe committed:
Merge branch 'main' into issue748

BSchelpe committed:
fix parameter change

BSchelpe committed:
restore arms parameter

BSchelpe committed:
Turn on the relevant test.

BSchelpe committed:
Change the type of error that is caught

BSchelpe committed:
Merge branch 'main' into issue748

smburdick committed:
Merge branch 'main' into issue748

smburdick commented:
> @smburdick, the test is now added

OK, I'll take a look tomorrow morning.

smburdick commented on tests/compile/compile_test.py:
```diff
@@ -447,7 +447,6 @@ def test_compile_spatial_hadamard_vertical_correlation_surface(
-@pytest.mark.skip(reason="Hadamard around spatial junction is not implemented yet.")
 @pytest.mark.parametrize(
     ("k", "convention", "direction", "obs_basis"),
     generate_inputs(CONVENTIONS, (Direction3D.X, Direction3D.Y), (Basis.X, Basis.Z)),
```
I assume we're not testing `Direction3D.Z` because we're not adding Hadamards in the temporal direction?

BSchelpe replied:
Yes, these are only the spatial hadamard tests.

smburdick commented on tests/compile/compile_test.py:
```diff
@@ -459,15 +458,27 @@ def test_compile_spatial_hadamard_horizontal_correlation_surface(
     d = 2 * k + 1
-    if convention.name == "fixed_bulk":
+    if convention.name == "fixed_boundary":
+        # This is not currently correctly implemented and throws an error complaining about
```
Can you make sure this case is documented where it should be thrown?

BSchelpe replied:
I've made an explicit NotImplementedError at the point in the code where it should be implemented. And changed this to catch NotImplemented.

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1106,6 +1158,114 @@ def _get_up_down_spatial_cube_arm_rpng_descriptions(
+        linked_cubes: tuple[CubeSpec, CubeSpec],
+        reset: Basis | None = None,
+        measurement: Basis | None = None,
+    ) -> Plaquettes:
```
I think this needs a docstring but I don't know the exact policy on private methods.

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -165,7 +166,9 @@ def get_3_body_rpng_descriptions(
-            convention: ``(top_left, top_right, bottom_left, bottom_right)``.
+            convention: ``(top_left, top_right, bottom_left, bottom_right)``. NB the colour of the
+            "corners" is pre-determined in the fixed bulk convention, so we do not need to send in a
+            basis to be chosen by the user.
```
"... so we don't need a user-defined basis"?

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1031,23 +1008,97 @@ def _get_left_right_spatial_cube_arm_rpng_descriptions(
         if SpatialArms.RIGHT in arms and _sbb == Basis.X and SpatialArms.UP in u.spatial_arms:
-            mapping[up] = _corners[1]
-        # Replaces the bottom plaquette if it should be a 3-body stabilizer.
+            del mapping[up]  # mapping[up] = _corners[1]
```
Please leave a comment explaining the commented out code, specific to each one.


https://stackoverflow.com/questions/3664288/is-it-bad-practice-to-leave-commented-out-code-in-production-releases

BSchelpe replied:
I left these commented out, because I wasn't sure deleting these corners was the best solution and I wanted to easily be able to reinstate them. I make a comment about this at the beginning of this code section on line 1018.

BSchelpe replied:
To expand on this: this part of the code needs to change relatively soon again after this PR to accommodate left-right hadamards. It is possible that the solution to implementing the left-right hadamards involves reinstating the commented out code, and so I wanted to leave it here to make it easy to reinstate.

BSchelpe replied:
Actually thinking some more. I don't think it is likely the solution will be to reinstate the commented out code. It would be cleaner to remove the comments, and they could always be found in the commit history if needed.

smburdick reviewed (commented):
minor comments.

BSchelpe committed:
Correct the horseshoe configuration

BSchelpe committed:
Merge branch 'issue748' of https://github.com/BSchelpe/tqec into issue748
Incorporate smburdick change.

BSchelpe committed:
Merge branch 'main' into issue748

BSchelpe committed:
smburdick's suggestions

BSchelpe committed:
Rationalise some naming

BSchelpe committed:
Missed one.

BSchelpe committed:
More name rationalisation

BSchelpe committed:
Improve docstrings

smburdick commented:
@BSchelpe is this ready for re-review?

BSchelpe commented:
> @BSchelpe is this ready for re-review?

Yes please. I believe I've addressed all the comments now. From my point of view I think this is finally ready to merge. @KabirDubey is there anything else you would like to see tested/changed?

smburdick commented on src/tqec/compile/specs/library/fixed_boundary.py:
```diff
@@ -295,7 +295,8 @@ def _get_spatial_cube_pipe_block(
         assert x is not None or y is not None
         spatial_boundary_basis: Basis = x if x is not None else y  # type: ignore
-        # Get the plaquette indices mappings
+        if spec.pipe_kind.has_hadamard:
```
Maybe leave a comment about why this isn't implemented yet.

BSchelpe replied:
There isn't a reason, beyond no one has done it yet.

smburdick commented on src/tqec/visualisation/computation/plaquette/extended.py:
```diff
@@ -35,10 +35,10 @@ def flip(self) -> ExtendedPlaquettePosition:
+    BOTTOM_RIGHT_TRIANGLE = auto()
+    RIGHT_HALF_RECTANGLE = auto()
+    TOP_LEFT_TRIANGLE = auto()
+    LEFT_HALF_RECTANGLE = auto()
```
Can you explain why these are the same, or are the old values being deprecated? Will there be other potential values added here?

Also, I would add a line to the commit when this merges that this enum is being swapped out, since it presumably affects other parts of the system.

BSchelpe replied:
The reason I've changed the names are they were tightly coupled to one use of the extended stabilisers, and the names were not applicable for other uses. These new names are purely based on the shape of the stabilisers and so are more general. I wouldn't have changed them as part of this PR, but I needed to add 2 new ones and the old naming convention could not be extended to them. I can't think of any other values that would want to be added in the future. I have searched for uses of this enum in the codebase and changed all instances.

smburdick commented on tests/compile/compile_test.py:
There's no code coverage report, but, I might add just another few tests, especially if it's around 80%.

BSchelpe replied:
I think it's best to add more tests separately from this PR, as the PR is already pretty big.

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1384,3 +1568,614 @@ def get_spatial_horizontal_hadamard_plaquettes(
             top_left_is_z_stabilizer, reset, measurement
         )
+
+    ###############################################################
```
Can you add more tests specifically under `library/` and `library/generators`?

BSchelpe replied:
I think it's best to add more tests separately from this PR, as the PR is already pretty big.

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1106,6 +1156,140 @@ def _get_up_down_spatial_cube_arm_rpng_descriptions(
+        """
+        u, v = linked_cubes
+        _sbb = spatial_boundary_basis
+        _sbb_f = _sbb.flipped()
```
variable `_sbb_f` is not used.

BSchelpe replied:
oops. Thanks. I've removed it.

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1106,6 +1156,140 @@ def _get_up_down_spatial_cube_arm_rpng_descriptions(
+        _sbb_f = _sbb.flipped()
+        # Note that in the below names, 'above' refers to above in the flat plane when the
+        # y-axis is pointing downwards. Ie 'above' refers to lower values of y.
+        if _sbb == Basis.Z:
```
I can't help but feel there's a more elegant (and less error prone) way to implement these if-else statements using a `dict`, but I could be wrong.

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1106,6 +1156,140 @@ def _get_up_down_spatial_cube_arm_rpng_descriptions(
+            elif (SpatialArms.RIGHT in v.spatial_arms) and (
+                SpatialArms.RIGHT not in u.spatial_arms
+            ):
+                plqts = self.get_spatial_z_above_rght_arm_extended_stabiliser_hadamard_plqts(
```
Why not just return `plqts` right after the assignment?

Ah, I made a comment about this 2 months ago. I think it should fit on the line if you return right after the assignment. Just makes it less error-prone.

BSchelpe replied:
It wasn't the length of line that was problematic in your suggested change. It was the lint error of too many return statements.

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1384,3 +1568,614 @@ def get_spatial_horizontal_hadamard_plaquettes(
+        reset: Basis | None = None,
+        measurement: Basis | None = None,
+    ) -> Plaquettes:
+        """Return a description of the plaquettes needed to implement a
```
set of the plaquettes?

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1384,3 +1568,614 @@ def get_spatial_horizontal_hadamard_plaquettes(
+        return Plaquettes(
+            FrozenDefaultDict(
+                {
+                    2: triangle[tlb].top,
```
I would add a comment or use some kind of data structure to illustrate what the range of indices means

BSchelpe replied:
Which indices are you referring to? Do you mean 2,4,5,6,7,8? Those are defined in the raw template, which the function's docstring refers to.

smburdick commented on src/tqec/compile/specs/library/generators/fixed_bulk.py:
```diff
@@ -1384,3 +1568,614 @@ def get_spatial_horizontal_hadamard_plaquettes(
+        """
+        return QubitHorizontalBorders()
+
+    def get_spatial_z_below_rght_arm_extended_stabiliser_hadamard_plqts(
```
Again, a lot of these methods feel a little redundant. For example:

`get_spatial_z_below_rght_arm_extended_stabiliser_hadamard_plqts`:
```
        tlb, otb = top_left_basis, top_left_basis.flipped()
        # Generating plaquette descriptions we will need later.
        extended_plaquette_collection = self.get_extended_plaquettes(reset, measurement)
        bulk = {
            tlb: extended_plaquette_collection[tlb].bulk,
            otb: extended_plaquette_collection[otb].bulk,
        }
        top_triangle = {
            tlb: extended_plaquette_collection[tlb].top_left_triangle,
            otb: extended_plaquette_collection[otb].top_left_triangle,
        }

        return Plaquettes(
            FrozenDefaultDict(
                {
                    2: top_triangle[tlb].top,
                    4: top_triangle[otb].bottom,
                    5: bulk[tlb].top,
                    6: bulk[otb].top,
                    7: bulk[otb].bottom,
                    8: bulk[tlb].bottom,
                }
            )
        )
```
vs
`get_spatial_x_no_lft_arms_extended_stabiliser_hadamard_plqts`
```
        # tlb: top-left basis, otb: other basis.
        tlb, otb = top_left_basis, top_left_basis.flipped()
        # Generating plaquette descriptions we will need later.
        extended_plaquette_collection = self.get_extended_plaquettes(reset, measurement)
        bulk = {
            tlb: extended_plaquette_collection[tlb].bulk,
            otb: extended_plaquette_collection[otb].bulk,
        }
        right_rectangle = {  # ie it is the right half of a 'normal' plaquette
            tlb: extended_plaquette_collection[tlb].right_half_rectangle,
            otb: extended_plaquette_collection[otb].right_half_rectangle,
        }
        return Plaquettes(
            FrozenDefaultDict(
                {
                    1: right_rectangle[tlb].top,
                    3: right_rectangle[otb].bottom,
                    5: bulk[otb].top,
                    6: bulk[tlb].top,
                    7: bulk[tlb].bottom,
                    8: bulk[otb].bottom,
                }
            )
        )
```

I recommend keeping them separate methods, but using something like a configuration dict to grab the specific triangle and indices that's specific to that method. That will save a lot of code, and make it a lot clearer and easier to extend.

smburdick reviewed (changes requested):
* I do think the methods in `fixed_bulk` can be generalized quite a bit, using a configuration, which would save a lot of code.
* Double check that the tests provide good coverage.

BSchelpe committed:
Merge branch 'main' into issue748

smburdick reviewed (dismissed):
Looks good. Thank you for addressing my comments
* We can add an action item for more tests if needed (maybe check code coverage again. I think it only shows up on the PR summary if you have write access due to the permissions settings)
* I think stylistically it is fine. We could probably do more as a team to specifically define what our coding standards should be, but it's consistent with the rest of the code.

BSchelpe committed:
small changes

BSchelpe commented:
> Looks good. Thank you for addressing my comments

Thank you! I've been thinking about your comments about changing to a configuration dict to reduce the amount of repeated code. I think I see how to do it, but it would be nice to do it in its own PR since the rearrangement is a bit complicated, and the code is currently functional so it would be nice to push this PR now. Thank you for approving! (Would you be able to reapprove? Sorry, I've just pushed the two small changes you suggested in your last review.)

smburdick committed:
Merge branch 'main' into issue748

smburdick reviewed (approved):
LGTM.

smburdick commented:
> > Looks good. Thank you for addressing my comments
> 
> Thank you! I've been thinking about your comments about changing to a configuration dict to reduce the amount of repeated code. I think I see how to do it, but it would be nice to do it in its own PR since the rearrangement is a bit complicated, and the code is currently functional so it would be nice to push this PR now. Thank you for approving! (Would you be able to reapprove? Sorry, I've just pushed the two small changes you suggested in your last review.)

Sure, I will raise a couple issues on these points.
